video/out/hwdec/dmabuf_interop_gl.c

```diff
@@ -160,46 +184,60 @@ static void vaapi_gl_mapper_uninit(const struct ra_hwdec_mapper *mapper)
 static bool vaapi_gl_map(struct ra_hwdec_mapper *mapper, bool probing)
 {
     struct priv *p_mapper = mapper->priv;
-    struct priv_owner *p_owner = mapper->owner->priv;
+    struct dmabuf_interop *p_owner = mapper->owner->priv;
     struct vaapi_gl_mapper_priv *p = p_mapper->interop_mapper_priv;
 
     GL *gl = ra_gl_get(mapper->ra);
 
-    for (int n = 0; n < p_mapper->num_planes; n++) {
-        if (p_mapper->desc.layers[n].num_planes > 1) {
-            // Should never happen because we request separate layers
-            MP_ERR(mapper, "Multi-plane VA surfaces are not supported\n");
-            return false;
-        }
+    for (int i = 0, n = 0; i < p_mapper->desc.nb_layers; i++) {
+        /*
+         * As we must map surfaces as one texture per plane, we can only support
+         * a subset of possible multi-plane layer formats. This is due to having
+         * to manually establish what DRM format each synthetic layer should
+         * have.
+         */
+        uint32_t format[AV_DRM_MAX_PLANES] = {
+            p_mapper->desc.layers[i].format,
+        };
 
-        int attribs[48] = {EGL_NONE};
-        int num_attribs = 0;
-
-        ADD_ATTRIB(EGL_LINUX_DRM_FOURCC_EXT, p_mapper->desc.layers[n].drm_format);
-        ADD_ATTRIB(EGL_WIDTH,  p_mapper->tex[n]->params.w);
-        ADD_ATTRIB(EGL_HEIGHT, p_mapper->tex[n]->params.h);
-
-        ADD_PLANE_ATTRIBS(0);
-        if (p_mapper->desc.layers[n].num_planes > 1)
-            ADD_PLANE_ATTRIBS(1);
-        if (p_mapper->desc.layers[n].num_planes > 2)
-            ADD_PLANE_ATTRIBS(2);
-        if (p_mapper->desc.layers[n].num_planes > 3)
-            ADD_PLANE_ATTRIBS(3);
-
-        p->images[n] = p->CreateImageKHR(eglGetCurrentDisplay(),
-            EGL_NO_CONTEXT, EGL_LINUX_DMA_BUF_EXT, NULL, attribs);
-        if (!p->images[n]) {
-            mp_msg(mapper->log, p_owner->probing_formats ? MSGL_DEBUG : MSGL_ERR,
-                   "Failed to import surface in EGL: %u\n", eglGetError());
-            return false;
+        if (p_mapper->desc.layers[i].nb_planes > 1) {
+            switch (p_mapper->desc.layers[i].format) {
+            case DRM_FORMAT_NV12:
+                format[0] = DRM_FORMAT_R8;
+                format[1] = DRM_FORMAT_GR88;
+                break;
+            default:
+                mp_msg(mapper->log, probing ? MSGL_DEBUG : MSGL_ERR,
+                       "Cannot map unknown multi-plane format: 0x%08X\n",
+                       p_mapper->desc.layers[i].format);
+                return false;
+            }
```

Comment:
might be premature but this could be moved into a lookup table

Reply:
Note: It would have to be iterated over, not statically indexed, because `DRM_FORMAT_NV12` is a fourcc (= large integer).

I'm not sure if that's an improvement over just a simple switch/case.